lib/main.js

```diff
@@ -0,0 +1,290 @@
+export function EventEmitter() {
+  this._eventListeners = new Map();
+  this._onceListeners = new Map();
+  this._maxListeners = 10;
+  this.on('maxListenersPassed', printMaxListenersWarning.bind(this));
+}
+
+// ## emitSync
+// Operates just like ``emit`` except all listeners must be synchronous.
+EventEmitter.prototype.emitSync = function emitSync(eventName, ...args) {
+  // Short-circuit for emit calls with no listeners
+  if(!this._onceListeners.has(eventName) &&
+    !this._eventListeners.has(eventName)) {
+    return;
+  }
+
+  // Copy listeners arrays to ensure listener call order
+  const onceListeners = [...(this._onceListeners.get(eventName) || [])];
+  const eventListeners = [...(this._eventListeners.get(eventName) || [])];
+
+  // process once listeners first
+  if(onceListeners.length > 0) {
+    for(const listener of onceListeners) {
+      // remove from emitter list
+      // get again in case already removed
+      const _onceListeners = this._onceListeners.get(eventName);
+      if(_onceListeners) {
+        const idx = _onceListeners.indexOf(listener);
+        if(idx !== -1) {
+          _onceListeners.splice(idx, 1);
+        }
+      }
+      // emit 'removeListener'
+      this.emitSync('removeListener', eventName, listener);
+
+      // invoke listener, exceptions are not caught
+      const result = listener(...args);
+      // check for async result
+      if(isPromise(result)) {
+        throw new TypeError(`async listeners not allowed for "${eventName}"`);
+      }
+      // check if event canceled
+      if(result === false) {
+        return false;
+      }
+    }
+    // remove once list if empty
+    const _onceListeners = this._onceListeners.get(eventName);
+    if(_onceListeners.length === 0) {
+      this._onceListeners.delete(eventName);
+    }
+  }
+
+  // process regular listeners
+  for(const listener of eventListeners) {
+    // invoke listener, exceptions are not caught
+    const result = listener(...args);
+    // check for async result
+    if(isPromise(result)) {
+      throw new TypeError(`async listeners not allowed for "${eventName}"`);
+    }
+    // check if event canceled
+    if(result === false) {
+      return false;
+    }
+  }
+};
+
+// ## emit
+// Takes and event name and a variable number of arguments. Listeners can be
+// asynchronous and return a Promise or synchronous. Listeners can cancel
+// events if their return value is `false` or is a Promise that resolves to
+// `false`. Event cancelation immediately ends all processing of event
+// listeners and returns `false`. If the event was emitted to all listeners
+// then undefined is returned. The order of event listeners is the order they
+// were attached, but ``once`` listeners are given priority over ``on``
+// listeners (so its possible to do a one-time cancelation and not have any of
+// the ``on`` listeners called).
+EventEmitter.prototype.emit = async function emit(eventName, ...args) {
```

Comment:
Is there any good reason to consider moving to a class?

Reply:
Yes, it should be converted to a class.  I was just working with getting the code running as is with newer semantics first.  Figured that could be put off until later if needed.

Reply:
Tracking for future: https://github.com/digitalbazaar/async-node-events/issues/7